### PR TITLE
VSphere Instance NodeAddresses ExternalID and InstanceID

### PR DESCRIPTION
**Pull request by dagnello**
This patch includes implementation for the following Instance object
interfaces:
- NodeAddresses
- ExternalID
- InstanceID

Also minor refactoring in overall Instance implementation.

### CODE REVIEW

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 1]
```diff
@@ -100,19 +103,49 @@ func vsphereLogin(cfg *VSphereConfig, ctx context.Context) (*govmomi.Client, err
...
 }
 
-type Instances struct {
-	cfg *VSphereConfig
+func getVirtualMachineByName(cfg *VSphereConfig, ctx context.Context, c *govmomi.Client, name string) (*object.VirtualMachine, error){
+	// Create a new finder
+	f := find.NewFinder(c.Client, true)
+
+	// Fetch and set data center
+	dc, err := f.DatacenterOrDefault(ctx, cfg.Global.Datacenter)
```
Comment (abithap) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: this should be Datacenter(ctx, cfg.Global.Datacenter) , DatacenterOrDefault() returns default dc if its unable to find the given datacenter.

Comment (dagnello) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: good catch, we should error if the configured datacenter is not valid

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -181,16 +258,79 @@ func (i *Instances) CurrentNodeName(hostname string) (string, error) {
...
+	err = getVirtualMachineManagedObjectReference(ctx, c, vm, "summary", &mvm)
+	if err !=nil {
+		return "", err
+	}
+
+	if mvm.Summary.Runtime.PowerState == ActivePowerState {
+		return vm.InventoryPath, nil
+	} 
+    
+    if mvm.Summary.Config.Template == false {
```
Comment (vipulsabhaya) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: tab vs. space issue

Comment (dagnello) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: fixed

**Pull request merged**